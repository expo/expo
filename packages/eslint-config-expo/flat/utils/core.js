const expo = require('eslint-plugin-expo');
const importPlugin = require('eslint-plugin-import');
const { jsExtensions } = require('./extensions');

const globals = require('globals');

module.exports = [
  importPlugin.flatConfigs.recommended,
  importPlugin.flatConfigs.errors,
  {
    plugins: {
      expo,
    },

    languageOptions: {
      globals: {
        console: 'readonly',
        exports: false,
        global: false,
        module: false,
        require: false,
      },

      ecmaVersion: 2022,
      sourceType: 'module',

      parserOptions: {
        ecmaFeatures: {
          impliedStrict: true,
          jsx: true,
        },
      },
    },

    settings: {
      'import/ignore': ['node_modules[\\\\/]+@?react-native'],

      'import/extensions': jsExtensions,
      'import/resolver': {
        node: { extensions: jsExtensions },
        typescript: true,
      },
    },

    rules: {
      eqeqeq: ['warn', 'smart'],
      'no-dupe-args': 'error',
      'no-dupe-class-members': 'error',
      'no-dupe-keys': 'error',
      'no-duplicate-case': 'error',
      'no-empty-character-class': 'warn',
      'no-empty-pattern': 'warn',
      'no-extend-native': 'warn',
      'no-extra-bind': 'warn',
      'no-redeclare': 'warn',
      'no-undef': 'error',
      'no-unreachable': 'warn',
      'no-unsafe-negation': 'warn',

      'no-unused-expressions': [
        'warn',
        {
          allowShortCircuit: true,
          enforceForJSX: true,
        },
      ],

      'no-unused-labels': 'warn',

      'no-unused-vars': [
        'warn',
        {
          vars: 'all',
          args: 'none',
          ignoreRestSiblings: true,
          caughtErrors: 'all',
          caughtErrorsIgnorePattern: '^_',
        },
      ],

      'no-with': 'warn',
      'unicode-bom': ['warn', 'never'],
      'use-isnan': 'error',
      'valid-typeof': 'error',
      'import/first': 'warn',
      'import/default': 'off',
    },
  },
  {
    files: ['**/*.d.ts'],

    rules: {
      'import/order': 'off',
    },
  },
  {
    files: ['**/metro.config.js'],

    languageOptions: {
      globals: globals.node,
    },
  },
];
