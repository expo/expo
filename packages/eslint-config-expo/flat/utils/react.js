const pluginReact = require('eslint-plugin-react');
const pluginReactHooks = require('eslint-plugin-react-hooks');

module.exports = [
  {
    plugins: {
      react: pluginReact,
      'react-hooks': pluginReactHooks,
    },

    languageOptions: {
      parserOptions: {
        ecmaFeatures: {
          jsx: true,
        },
      },
    },

    settings: {
      react: {
        version: 'detect',
      },
    },

    rules: {
      ...pluginReactHooks.configs.recommended.rules,
      'react/display-name': 'warn',
      'react/jsx-no-duplicate-props': 'error',
      'react/jsx-no-undef': 'error',
      'react/jsx-uses-react': 'warn',
      'react/jsx-uses-vars': 'warn',
      'react/no-danger-with-children': 'warn',
      'react/no-deprecated': 'warn',
      'react/no-direct-mutation-state': 'warn',

      'react/no-string-refs': [
        'warn',
        {
          noTemplateLiterals: true,
        },
      ],

      'react/no-this-in-sfc': 'warn',
      'react/no-unknown-property': 'warn',
      'react/require-render-return': 'warn',
    },
  },
];
