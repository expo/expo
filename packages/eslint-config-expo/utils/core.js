const { jsExtensions } = require('./extensions');

module.exports = {
  parserOptions: {
    sourceType: 'module',
    ecmaVersion: 2022,
    ecmaFeatures: { impliedStrict: true, jsx: true },
  },
  env: { es2022: true },
  globals: {
    console: 'readonly',
    exports: false,
    global: false,
    module: false,
    require: false,
  },
  extends: ['plugin:import/errors'],
  plugins: ['import', 'expo'],
  rules: {
    eqeqeq: ['warn', 'smart'],
    'no-dupe-args': 'error',
    'no-dupe-class-members': 'error',
    'no-dupe-keys': 'error',
    'no-duplicate-case': 'error',
    'no-empty-character-class': 'warn',
    'no-empty-pattern': 'warn',
    'no-extend-native': 'warn',
    'no-extra-bind': 'warn',
    'no-redeclare': 'warn',
    'no-undef': 'error',
    'no-unreachable': 'warn',
    'no-unsafe-negation': 'warn',
    'no-unused-expressions': ['warn', { allowShortCircuit: true, enforceForJSX: true }],
    'no-unused-labels': 'warn',
    'no-unused-vars': [
      'warn',
      {
        vars: 'all',
        args: 'none',
        ignoreRestSiblings: true,
        caughtErrors: 'all',
        caughtErrorsIgnorePattern: '^_',
      },
    ],
    'no-with': 'warn',
    'unicode-bom': ['warn', 'never'],
    'use-isnan': 'error',
    'valid-typeof': 'error',
    'import/first': 'warn', //keep
    'import/default': 'off',
  },
  settings: {
    'import/extensions': jsExtensions,
    'import/ignore': [
      // react-native's main module is Flow, not JavaScript, and raises parse errors. Additionally,
      // several other react-native-related packages still publish Flow code as their main source.
      'node_modules[\\\\/]+@?react-native',
    ],
    'import/resolver': {
      node: { extensions: jsExtensions },
      typescript: true
    },
  },
  overrides: [
    {
      files: ['*.d.ts'],
      rules: {
        'import/order': 'off',
      },
    },
  ],
};
