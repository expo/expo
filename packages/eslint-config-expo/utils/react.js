module.exports = {
  parserOptions: { ecmaFeatures: { jsx: true } },
  plugins: ['react'],
  extends: ['plugin:react-hooks/recommended'],
  rules: {
    'react/display-name': 'warn',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-undef': 'error',
    'react/jsx-uses-react': 'warn',
    'react/jsx-uses-vars': 'warn',
    'react/no-danger-with-children': 'warn',
    'react/no-deprecated': 'warn',
    'react/no-direct-mutation-state': 'warn',
    'react/no-string-refs': ['warn', { noTemplateLiterals: true }],
    'react/no-this-in-sfc': 'warn',
    'react/no-unknown-property': 'warn',
    'react/require-render-return': 'warn',
  },
  settings: {
    react: { version: 'detect' },
  },
};
