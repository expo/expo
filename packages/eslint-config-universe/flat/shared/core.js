const { defineConfig } = require('eslint/config');
const importPlugin = require('eslint-plugin-import');
const nodePlugin = require('eslint-plugin-n');
const globals = require('globals');

const { jsExtensions } = require('./extensions');

module.exports = defineConfig([
  importPlugin.flatConfigs.recommended,
  {
    plugins: {
      n: nodePlugin,
    },

    languageOptions: {
      globals: {
        ...globals.jest,
        console: false,
        exports: false,
        global: false,
        module: false,
        require: false,
      },

      ecmaVersion: 2022,
      sourceType: 'module',

      parserOptions: {
        ecmaFeatures: {
          impliedStrict: true,
          jsx: true,
        },
      },
    },

    settings: {
      'import/extensions': jsExtensions,
      'import/ignore': ['node_modules[\\\\/]+@?react-native'],

      'import/resolver': {
        node: {
          extensions: jsExtensions,
        },
      },
    },

    rules: {
      'array-bracket-spacing': ['warn', 'never'],

      'arrow-spacing': [
        'warn',
        {
          before: true,
          after: true,
        },
      ],

      curly: ['warn', 'all'],
      'block-spacing': ['warn', 'always'],
      'comma-dangle': ['warn', 'always-multiline'],

      'comma-spacing': [
        'warn',
        {
          before: false,
          after: true,
        },
      ],

      'comma-style': ['warn', 'last'],
      'computed-property-spacing': ['warn', 'never'],
      'constructor-super': 'warn',
      'dot-location': ['warn', 'property'],
      'eol-last': 'warn',
      eqeqeq: ['warn', 'smart'],
      'func-call-spacing': ['warn', 'never'],
      'generator-star-spacing': ['warn', 'after'],
      'getter-return': 'warn',
      'jsx-quotes': ['warn', 'prefer-double'],
      'new-parens': 'warn',
      'no-alert': 'off',
      'no-array-constructor': 'warn',
      'no-caller': 'warn',
      'no-case-declarations': 'warn',
      'no-compare-neg-zero': 'warn',
      'no-cond-assign': 'warn',
      'no-const-assign': 'error',

      'no-constant-condition': [
        'warn',
        {
          checkLoops: false,
        },
      ],

      'no-control-regex': 'off',
      'no-debugger': 'warn',
      'no-delete-var': 'error',
      'no-dupe-args': 'error',
      'no-dupe-class-members': 'error',
      'no-dupe-keys': 'error',
      'no-duplicate-case': 'error',
      'no-empty-character-class': 'warn',
      'no-empty-pattern': 'warn',
      'no-eval': 'warn',
      'no-ex-assign': 'warn',
      'no-extend-native': 'warn',
      'no-extra-bind': 'warn',
      'no-extra-boolean-cast': 'warn',
      'no-extra-semi': 'warn',
      'no-fallthrough': 'warn',
      'no-floating-decimal': 'warn',
      'no-func-assign': 'error',
      'no-global-assign': 'warn',
      'no-implied-eval': 'warn',
      'no-inner-declarations': 'warn',
      'no-invalid-regexp': 'error',
      'no-irregular-whitespace': 'warn',
      'no-iterator': 'warn',
      'no-label-var': 'warn',

      'no-labels': [
        'warn',
        {
          allowLoop: true,
          allowSwitch: true,
        },
      ],

      'no-lone-blocks': 'warn',
      'no-multi-assign': 'warn',
      'no-new': 'warn',
      'no-new-func': 'warn',
      'no-object-constructor': 'warn',
      'no-new-native-nonconstructor': 'error',
      'no-obj-calls': 'warn',
      'no-octal': 'warn',
      'no-octal-escape': 'warn',
      'no-mixed-spaces-and-tabs': 'warn',
      'no-proto': 'warn',
      'no-redeclare': 'warn',
      'no-return-assign': 'warn',
      'no-script-url': 'warn',
      'no-self-assign': 'warn',
      'no-self-compare': 'warn',
      'no-sequences': 'warn',
      'no-shadow-restricted-names': 'warn',
      'no-sparse-arrays': 'warn',
      'no-this-before-super': 'warn',
      'no-throw-literal': 'warn',
      'no-trailing-spaces': 'warn',
      'no-undef': 'error',
      'no-unexpected-multiline': 'warn',
      'no-unneeded-ternary': 'warn',
      'no-unreachable': 'warn',
      'no-unsafe-negation': 'warn',

      'no-unused-expressions': [
        'warn',
        {
          allowShortCircuit: true,
          enforceForJSX: true,
        },
      ],

      'no-unused-labels': 'warn',

      'no-unused-vars': [
        'warn',
        {
          vars: 'all',
          args: 'none',
          ignoreRestSiblings: true,
          caughtErrors: 'all',
          caughtErrorsIgnorePattern: '^_',
        },
      ],

      'no-useless-computed-key': 'warn',
      'no-useless-concat': 'warn',
      'no-useless-constructor': 'warn',
      'no-useless-escape': 'warn',
      'no-useless-rename': 'warn',
      'no-useless-return': 'warn',
      'no-var': 'warn',
      'no-void': 'warn',
      'no-whitespace-before-property': 'warn',
      'no-with': 'warn',
      'object-shorthand': 'warn',

      'operator-linebreak': [
        'warn',
        'after',
        {
          overrides: {
            '?': 'before',
            ':': 'before',
          },
        },
      ],

      'prefer-const': [
        'warn',
        {
          destructuring: 'all',
          ignoreReadBeforeAssign: true,
        },
      ],

      'prefer-promise-reject-errors': 'warn',
      'prefer-rest-params': 'warn',
      'prefer-spread': 'warn',
      radix: 'warn',
      'rest-spread-spacing': ['warn', 'never'],
      semi: 'warn',

      'semi-spacing': [
        'warn',
        {
          before: false,
          after: true,
        },
      ],

      'semi-style': ['warn', 'last'],
      'space-before-blocks': ['warn', 'always'],

      'space-before-function-paren': [
        'warn',
        {
          anonymous: 'never',
          named: 'never',
        },
      ],

      'space-in-parens': ['warn', 'never'],
      'space-infix-ops': 'warn',

      'switch-colon-spacing': [
        'warn',
        {
          before: false,
          after: true,
        },
      ],

      'template-curly-spacing': ['warn', 'never'],
      'template-tag-spacing': ['warn', 'never'],
      'unicode-bom': ['warn', 'never'],
      'use-isnan': 'error',
      'valid-typeof': 'error',
      'yield-star-spacing': ['warn', 'after'],

      yoda: [
        'warn',
        'never',
        {
          exceptRange: true,
        },
      ],

      'import/default': 'off',
      'import/export': 'error',
      'import/first': 'warn',

      'import/namespace': [
        'error',
        {
          allowComputed: true,
        },
      ],

      'import/no-duplicates': 'error',

      // TODO (Kadi): Enable this. Disabling for now because import/recommended adds it, but we didn't use to have it enabled
      'import/no-unresolved': 'off',

      'import/order': [
        'warn',
        {
          groups: [['builtin', 'external'], 'internal', ['parent', 'index', 'sibling']],
          'newlines-between': 'always',

          alphabetize: {
            order: 'asc',
          },
        },
      ],

      'n/handle-callback-err': ['warn', '^(e|err|error|.+Error)$'],
      'n/no-new-require': 'warn',
    },
  },
  {
    files: ['**/*.d.ts'],

    rules: {
      'import/order': 'off',
    },
  },
]);
