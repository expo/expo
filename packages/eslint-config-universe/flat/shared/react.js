const { defineConfig } = require('eslint/config');
const react = require('eslint-plugin-react');
const hooksPlugin = require('eslint-plugin-react-hooks');

module.exports = defineConfig([
  {
    plugins: {
      react,
      'react-hooks': hooksPlugin,
    },

    languageOptions: {
      // TODO: Const is a reserved word
      // ecmaVersion: 5,
      // sourceType: 'script',

      parserOptions: {
        ecmaFeatures: {
          jsx: true,
        },
      },
    },

    settings: {
      react: {
        version: 'detect',
      },
    },

    rules: {
      ...hooksPlugin.configs.recommended.rules,
      'react/jsx-boolean-value': ['warn', 'never'],

      'react/jsx-closing-bracket-location': [
        'warn',
        {
          nonEmpty: 'after-props',
          selfClosing: 'tag-aligned',
        },
      ],

      'react/jsx-curly-brace-presence': ['warn', 'never'],

      'react/jsx-curly-spacing': [
        'warn',
        {
          when: 'never',
        },
      ],

      'react/jsx-equals-spacing': ['warn', 'never'],
      'react/jsx-first-prop-new-line': ['warn', 'multiline'],
      'react/jsx-fragments': ['warn', 'syntax'],
      'react/jsx-indent': ['warn', 2],
      'react/jsx-indent-props': ['warn', 2],

      'react/jsx-no-bind': [
        'warn',
        {
          allowArrowFunctions: true,
          allowFunctions: true,
        },
      ],

      'react/jsx-no-duplicate-props': 'error',
      'react/jsx-no-undef': 'error',
      'react/jsx-props-no-multi-spaces': 'warn',
      'react/jsx-tag-spacing': 'warn',
      'react/jsx-uses-react': 'warn',
      'react/jsx-uses-vars': 'warn',

      'react/jsx-wrap-multilines': [
        'warn',
        {
          declaration: false,
          assignment: false,
          return: true,
          arrow: true,
        },
      ],

      'react/no-access-state-in-setstate': 'warn',
      'react/no-did-mount-set-state': 'warn',
      'react/no-did-update-set-state': 'warn',
      'react/no-direct-mutation-state': 'warn',
      'react/no-redundant-should-component-update': 'warn',
      'react/no-this-in-sfc': 'warn',
      'react/no-unknown-property': 'warn',
      'react/no-will-update-set-state': 'warn',
      'react/require-render-return': 'warn',
      'react/self-closing-comp': 'warn',
      'react-hooks/rules-of-hooks': 'error',
      'react-hooks/exhaustive-deps': 'off',
    },
  },
]);
