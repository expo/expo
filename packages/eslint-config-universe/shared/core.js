const { jsExtensions } = require('./extensions');

module.exports = {
  parserOptions: {
    sourceType: 'module',
    ecmaVersion: 2022,
    ecmaFeatures: { impliedStrict: true, jsx: true },
  },
  env: { es2022: true, jest: true },
  globals: {
    console: false,
    exports: false,
    global: false,
    module: false,
    require: false,
  },
  plugins: ['import', 'node'],
  rules: {
    'array-bracket-spacing': ['warn', 'never'],
    'arrow-spacing': ['warn', { before: true, after: true }],
    curly: ['warn', 'all'],
    'block-spacing': ['warn', 'always'],
    'comma-dangle': ['warn', 'always-multiline'],
    'comma-spacing': ['warn', { before: false, after: true }],
    'comma-style': ['warn', 'last'],
    'computed-property-spacing': ['warn', 'never'],
    'constructor-super': 'warn',
    'dot-location': ['warn', 'property'],
    'eol-last': 'warn',
    eqeqeq: ['warn', 'smart'],
    'func-call-spacing': ['warn', 'never'],
    'generator-star-spacing': ['warn', 'after'],
    'getter-return': 'warn',
    'jsx-quotes': ['warn', 'prefer-double'],
    'new-parens': 'warn',
    'no-alert': 'off',
    'no-array-constructor': 'warn',
    'no-caller': 'warn',
    'no-case-declarations': 'warn',
    'no-compare-neg-zero': 'warn',
    'no-cond-assign': 'warn',
    'no-const-assign': 'error',
    'no-constant-condition': ['warn', { checkLoops: false }],
    'no-control-regex': 'off',
    'no-debugger': 'warn',
    'no-delete-var': 'error',
    'no-dupe-args': 'error',
    'no-dupe-class-members': 'error',
    'no-dupe-keys': 'error',
    'no-duplicate-case': 'error',
    'no-empty-character-class': 'warn',
    'no-empty-pattern': 'warn',
    'no-eval': 'warn',
    'no-ex-assign': 'warn',
    'no-extend-native': 'warn',
    'no-extra-bind': 'warn',
    'no-extra-boolean-cast': 'warn',
    'no-extra-semi': 'warn',
    'no-fallthrough': 'warn',
    'no-floating-decimal': 'warn',
    'no-func-assign': 'error',
    'no-global-assign': 'warn',
    'no-implied-eval': 'warn',
    'no-inner-declarations': 'warn',
    'no-invalid-regexp': 'error',
    'no-irregular-whitespace': 'warn',
    'no-iterator': 'warn',
    'no-label-var': 'warn',
    'no-labels': ['warn', { allowLoop: true, allowSwitch: true }],
    'no-lone-blocks': 'warn',
    'no-multi-assign': 'warn',
    'no-new': 'warn',
    'no-new-func': 'warn',
    'no-new-object': 'warn',
    'no-new-symbol': 'error',
    'no-obj-calls': 'warn',
    'no-octal': 'warn',
    'no-octal-escape': 'warn',
    'no-mixed-spaces-and-tabs': 'warn',
    'no-proto': 'warn',
    'no-redeclare': 'warn',
    'no-return-assign': 'warn',
    'no-script-url': 'warn',
    'no-self-assign': 'warn',
    'no-self-compare': 'warn',
    'no-sequences': 'warn',
    'no-shadow-restricted-names': 'warn',
    'no-sparse-arrays': 'warn',
    'no-this-before-super': 'warn',
    'no-throw-literal': 'warn',
    'no-trailing-spaces': 'warn',
    'no-undef': 'error',
    'no-unexpected-multiline': 'warn',
    'no-unneeded-ternary': 'warn',
    'no-unreachable': 'warn',
    'no-unsafe-negation': 'warn',
    'no-unused-expressions': ['warn', { allowShortCircuit: true, enforceForJSX: true }],
    'no-unused-labels': 'warn',
    'no-unused-vars': [
      'warn',
      {
        vars: 'all',
        args: 'none',
        ignoreRestSiblings: true,
        caughtErrors: 'all',
        caughtErrorsIgnorePattern: '^_',
      },
    ],
    'no-useless-computed-key': 'warn',
    'no-useless-concat': 'warn',
    'no-useless-constructor': 'warn',
    'no-useless-escape': 'warn',
    'no-useless-rename': 'warn',
    'no-useless-return': 'warn',
    'no-var': 'warn',
    'no-void': 'warn',
    'no-whitespace-before-property': 'warn',
    'no-with': 'warn',
    'object-shorthand': 'warn',
    'operator-linebreak': ['warn', 'after', { overrides: { '?': 'before', ':': 'before' } }],
    'prefer-const': ['warn', { destructuring: 'all', ignoreReadBeforeAssign: true }],
    'prefer-promise-reject-errors': 'warn',
    'prefer-rest-params': 'warn',
    'prefer-spread': 'warn',
    radix: 'warn',
    'rest-spread-spacing': ['warn', 'never'],
    semi: 'warn',
    'semi-spacing': ['warn', { before: false, after: true }],
    'semi-style': ['warn', 'last'],
    'space-before-blocks': ['warn', 'always'],
    'space-before-function-paren': ['warn', { anonymous: 'never', named: 'never' }],
    'space-in-parens': ['warn', 'never'],
    'space-infix-ops': 'warn',
    'switch-colon-spacing': ['warn', { before: false, after: true }],
    'template-curly-spacing': ['warn', 'never'],
    'template-tag-spacing': ['warn', 'never'],
    'unicode-bom': ['warn', 'never'],
    'use-isnan': 'error',
    'valid-typeof': 'error',
    'yield-star-spacing': ['warn', 'after'],
    yoda: ['warn', 'never', { exceptRange: true }],

    'import/default': 'off',
    'import/export': 'error',
    'import/first': 'warn',
    'import/namespace': ['error', { allowComputed: true }],
    'import/no-duplicates': 'error',
    'import/order': [
      'warn',
      {
        groups: [['builtin', 'external'], 'internal', ['parent', 'index', 'sibling']],
        'newlines-between': 'always',
        alphabetize: {
          order: 'asc',
        },
      },
    ],

    'node/handle-callback-err': ['warn', '^(e|err|error|.+Error)$'],
    'node/no-new-require': 'warn',
  },
  settings: {
    'import/extensions': jsExtensions,
    'import/ignore': [
      // react-native's main module is Flow, not JavaScript, and raises parse errors. Additionally,
      // several other react-native-related packages still publish Flow code as their main source.
      'node_modules[\\\\/]+@?react-native',
    ],
    'import/resolver': {
      node: { extensions: jsExtensions },
    },
  },
  overrides: [
    {
      files: ['*.d.ts'],
      rules: {
        'import/order': 'off',
      },
    },
  ],
};
