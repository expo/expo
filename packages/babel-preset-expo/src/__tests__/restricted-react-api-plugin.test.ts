/**
 * Copyright Â© 2024 650 Industries.
 */

import * as babel from '@babel/core';

import preset from '..';

const ENABLED_CALLER = {
  name: 'metro',
  isDev: false,
  isServer: false,
  projectRoot: '/',
};

function getCaller(props: Record<string, string | boolean>): babel.TransformCaller {
  return props as unknown as babel.TransformCaller;
}

const DEF_OPTIONS = {
  // Ensure this is absolute to prevent the filename from being converted to absolute and breaking CI tests.
  filename: '/unknown',

  babelrc: false,
  presets: [[preset, { disableImportExportTransform: true }]],
  sourceMaps: true,
  configFile: false,
  compact: false,
  comments: true,
  retainLines: false,
  caller: getCaller({ ...ENABLED_CALLER, platform: 'ios' }),
};

const originalEnv = process.env;

beforeEach(() => {
  process.env = { ...originalEnv, FORCE_COLOR: '0' };
});

afterAll(() => {
  process.env = { ...originalEnv };
});

function getOpts(caller: Record<string, string | boolean>) {
  return {
    ...DEF_OPTIONS,
    caller: getCaller({ ...ENABLED_CALLER, ...caller }),
  };
}

describe('forbidden React server APIs', () => {
  function runServerPass(src: string) {
    return babel.transform(
      src,
      getOpts({
        isReactServer: true,
        platform: 'ios',
      })
    );
  }
  it(`does not assert importing client-side APIs in client components (react server mode)`, () => {
    // This test covers the order of server registry running before the assertion to remove the import.
    expect(runServerPass(`"use client"; import { useState } from 'react';`).code).toMatch(
      'react-server-dom-webpack'
    );
  });
  it(`asserts importing client-side React APIs in server components`, () => {
    expect(() => runServerPass(`import { useState } from 'react';`)).toThrowErrorMatchingSnapshot();
    expect(() => runServerPass(`import { useRef, useContext } from 'react';`)).toThrowError();
    expect(() => runServerPass(`import React, { useRef } from 'react';`)).toThrowError();
    expect(() => runServerPass(`import { PureComponent } from 'react';`)).toThrowError();
    expect(() => runServerPass(`import { Component } from 'react';`)).toThrowError();
    expect(() => runServerPass(`import { useRandom } from 'react';`)).not.toThrowError();
  });
  it(`asserts importing client-side react-dom APIs in server components`, () => {
    expect(() =>
      runServerPass(`import { findDOMNode } from 'react-dom';`)
    ).toThrowErrorMatchingSnapshot();
    expect(() => runServerPass(`import { useRandom } from 'react-dom';`)).not.toThrowError();
  });
  it(`does not assert importing client-side react-dom APIs in server components if they are in node modules`, () => {
    expect(
      babel.transform(`import { findDOMNode } from 'react-dom';`, {
        ...DEF_OPTIONS,
        filename: '/bacon/node_modules/@bacons/breakfast.js',
        caller: getCaller({ ...ENABLED_CALLER, isReactServer: true, platform: 'ios' }),
      }).code
    ).toBe(`import { findDOMNode } from 'react-dom';`);
  });

  it(`asserts client-side React API usage in server components`, () => {
    expect(() =>
      runServerPass(`
    import * as React from 'react';
    
    export default function App() {
      const [state, setState] = React.useState(0);
      return <div>{state}</div>;
    }
    `)
    ).toThrow(/cannot be used in a React server component/);
  });
  it(`asserts client-side React API usage in server components (default import)`, () => {
    expect(() =>
      runServerPass(`
    import React from 'react';
    
    export default function App() {
      const ref = React.useRef(null);
      return <div>{ref}</div>;
    }
    `)
    ).toThrow(/cannot be used in a React server component/);
  });
  it(`asserts client-side React class component usage in server components`, () => {
    expect(() =>
      runServerPass(`
    import React from 'react';
    
    class App extends React.Component {
        render() {
        return <div />;
        }
    }
    `)
    ).toThrow(/Class components cannot be/);
  });

  it(`allows client-side React API usage in client components`, () => {
    runServerPass(`
    "use client"
    import React from 'react';
    
    export default function App() {
      const ref = React.useRef(null);
      return <div>{ref}</div>;
    }
    `);
    runServerPass(`
    "use client"
    import * as React from 'react';
    
    export default function App() {
      const ref = React.useRef(null);
      return <div>{ref}</div>;
    }
    `);
  });

  it(`allows client-side React class component usage in client components`, () => {
    runServerPass(`
      "use client"
    import React from 'react';
    
    class App extends React.Component {
        render() {
        return <div />;
        }
    }
    `);
  });
});
