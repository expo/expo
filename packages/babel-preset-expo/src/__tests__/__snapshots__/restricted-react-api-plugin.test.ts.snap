// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`forbidden React server APIs asserts importing client-side React APIs in server components 1`] = `
"/unknown: Client-only "react" API "useState" cannot be imported in a React server component. Add the "use client" directive to the top of this file or one of the parent files to enable running this stateful code on a user's device.
[0m[31m[1m>[22m[39m[90m 1 |[39m [36mimport[39m { useState } [36mfrom[39m [32m'react'[39m[33m;[39m[0m
[0m [90m   |[39m [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[0m"
`;

exports[`forbidden React server APIs asserts importing client-side react-dom APIs in server components 1`] = `
"/unknown: Client-only "react-dom" API "findDOMNode" cannot be imported in a React server component. Add the "use client" directive to the top of this file or one of the parent files to enable running this stateful code on a user's device.
[0m[31m[1m>[22m[39m[90m 1 |[39m [36mimport[39m { findDOMNode } [36mfrom[39m [32m'react-dom'[39m[33m;[39m[0m
[0m [90m   |[39m [31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[31m[1m^[22m[39m[0m"
`;
